bokeh/client/session.py

```diff
@@ -306,11 +323,23 @@ def _document_changed(self, event):
             return
 
         if isinstance(event, SessionCallbackAdded):
-            self._add_periodic_callback(event.callback)
+            if isinstance(event.callback, PeriodicCallback):
+                self._add_periodic_callback(event.callback)
+            elif isinstance(event.callback, TimeoutCallback):
+                self._add_timeout_callback(event.callback)
+            else:
+                raise ValueError("Expected callback of type PeriodicCallback or TimeoutCallback, got: %s" % event.callback)
+
             return
 
-        if isinstance(event, SessionCallbackRemoved):
-            self._remove_periodic_callback(event.callback)
+        elif isinstance(event, SessionCallbackRemoved):
+            if isinstance(event.callback, PeriodicCallback):
+                self._remove_periodic_callback(event.callback)
+            elif isinstance(event.callback, TimeoutCallback):
+                self._remove_timeout_callback(event.callback)
+            else:
+                raise ValueError("Expected callback of type PeriodicCallback or TimeoutCallback, got: %s" % event.callback)
+
```

Comment:
if you go the event.dispatch route you might want to make it its own PR since you'd have to do all the existing events as well as your new ones.